libtcmu.c

```diff
@@ -383,18 +383,192 @@ static void tcmu_cdb_debug_info(const struct tcmulib_cmd *cmd)
+                        }
+
+                        aio_entry = list_first_entry(&aio_wq->aio_queue,
+                                                     struct tcmu_aio_entry, entry);
+                        tcmulib_track_aio_request(dev, &aio_entry->entry);
+                }
+                pthread_mutex_unlock(&aio_wq->aio_lock);
+                pthread_cleanup_pop(0);
+
+                ret = aio_entry->call(dev, aio_entry->cmd);
+
+                tcmulib_command_complete(dev, aio_entry->cmd, ret);
+                if (tcmulib_can_wake_up_kernel(dev, &aio_entry->entry)) {
+                        tcmulib_processing_complete(dev);
+                }
+
+                free(aio_entry);
+        }
+
+        return NULL;
+}
+
+static int aio_schedule(struct tcmu_device *dev, struct tcmulib_cmd *cmd,
+                        int (*call) (struct tcmu_device *, struct tcmulib_cmd *))
+{
+        struct tcmu_aio_entry *aio_entry;
+        struct tcmu_aio_queue *aio_wq = &dev->work_queue;
+
+        aio_entry = malloc(sizeof(*aio_entry));
+        if (!aio_entry) {
+                return -ENOMEM;
+        }
+
+        list_node_init(&aio_entry->entry);
+
+        aio_entry->cmd = cmd;
+        aio_entry->call = call;
+
+        /* cleanup push/pop not _really_ required here atm */
+        pthread_cleanup_push(_cleanup_mutex_lock, (void *)&aio_wq->aio_lock);
+        pthread_mutex_lock(&aio_wq->aio_lock);
+        {
+                list_add_tail(&aio_wq->aio_queue, &aio_entry->entry);
+                pthread_cond_signal(&aio_wq->aio_cond); // TODO: conditional
+        }
+        pthread_mutex_unlock(&aio_wq->aio_lock);
+        pthread_cleanup_pop(0);
+
+        return TCMU_ASYNC_HANDLED;
+}
+
+static int setup_aio_work_queue(struct tcmu_device *dev)
+{
+        int ret;
+        struct tcmu_aio_queue *aio_wq = &dev->work_queue;
+
+        list_head_init(&aio_wq->aio_queue);
+
+        ret = pthread_mutex_init(&aio_wq->aio_lock, NULL);
+        if (ret < 0) {
+                goto out;
+        }
+        ret = pthread_cond_init(&aio_wq->aio_cond, NULL);
+        if (ret < 0) {
+                goto cleanup_lock;
+        }
+
+        // TODO: >1 worker threads?
+        ret = pthread_create(&aio_wq->aio_wq_thread, NULL, aio_work_queue, dev);
+        if (ret < 0) {
+                goto cleanup_cond;
+        }
+
+        return 0;
+
+ cleanup_cond:
+        pthread_cond_destroy(&aio_wq->aio_cond);
+ cleanup_lock:
+        pthread_mutex_destroy(&aio_wq->aio_lock);
+ out:
+        return ret;
+}
+
+static void cleanup_aio_work_queue(struct tcmu_device *dev,
+                                   bool cancel)
+{
+        int ret;
+        struct tcmu_aio_entry *curr, *next;
+        struct tcmu_aio_queue *aio_wq = &dev->work_queue;
+
+        if (cancel) {
+                cancel_thread(aio_wq->aio_wq_thread);
+        }
+
+        list_for_each_safe(&aio_wq->aio_queue, curr, next, entry) {
+                list_del(&curr->entry);
```

Comment:
Right, but I'm not sure what should be the return value. Also, thinking that `tcmulib_processing_complete()` need to be called as that seems unnecessary after device removal.